### DISCUSSION TITLE
Question : Must --simulate option simulate database pulling (mysqldump) ?

### DISCUSSION THREAD

**Discussion by eldk** [welaika/wordmove#629] [General]
Hello,

Must --simulate option simulate database pulling ?

```
wordmove pull --all -v --debug --simulate

>f..t...... wp-content/languages/themes/twentytwentyone-fr_FR.po
sent 853 bytes  received 6,625 bytes  4,985.33 bytes/sec
total size is 5,594,481  speedup is 748.13 (DRY RUN)

▬▬ Pulling Database ▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬
```
Nothing happened ?

The config use ssh + rsync + mysqldump.

Thanks,

Eric

PS : wordmove doctor checks are all green

**Comment by alessandro-fazzi** [comment]
Hi,

actually when simulating the pull action will solely print database actions, actually skipping them. Sincerly I've never realized my own expectation about what simulating the DB transfer/adapt could mean...

As it stands it shows to you only the operation that would have been ran, without running it.

If you have different expectations I'd like to discuss them and eventually track them down in a feature request.

Thanks for the enquiry.

**Comment by eldk** [comment]
Thanks for your answer.

>As it stands it shows to you only the operation that would have been ran, without running it.

So I should see log lines under -- Pulling Database --- ?

Those lines should reflect what should be done without --simulate option ?

I have nothing under (not a single line)  -- Pulling Database --- no database actions printed at all.

Thanks,

Eric

PS : I was using wordmove with sftp before switching to SSH, and, if I remember database operation were logged under -- Pulling Database when --simulate option is used (even if nothing is really done). 

```
wordmove --version
5.2.1
```

**Comment by alessandro-fazzi** [comment]
Yes, sorry: the "operation" from the wordmove pov is "Pulling database".

This is what i get with FTP

<img width="314" alt="immagine" src="https://user-images.githubusercontent.com/904690/148223637-3bf6521d-e472-42f2-b675-c1b3f476caa1.png">

While I'm working on the new wordmove version I'm compelling about what to log and what to not log while simulating. This is an example from the alpha version

<img width="746" alt="immagine" src="https://user-images.githubusercontent.com/904690/148223819-8c6d59c1-d4cf-49f6-99e1-27a896c05ca6.png">

But you didn't express any expectation about what you think should be useful. That would help with future implementation. :)

**Comment by eldk** [comment]
Hello,

Removing --simulate, I have some error message when pulling db :

`wordmove pull --all --verbose --debug`

```
 local | gzip -9 -f "[secret]/wp-content/local-backup-1641848292.sql"

   remote | mysqldump --host=[secret] --user=mysuser --password=[secret] --result-file="[secret]/wp-content/dump.sql" --verbose database
Traceback (most recent call last):
	32: from /usr/local/bin/wordmove:23:in `<main>'
	31: from /usr/local/bin/wordmove:23:in `load'
	30: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/exe/wordmove:6:in `<top (required)>'
	29: from /var/lib/gems/2.7.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	28: from /var/lib/gems/2.7.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	27: from /var/lib/gems/2.7.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	26: from /var/lib/gems/2.7.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	25: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/cli.rb:91:in `pull'
	24: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/cli.rb:39:in `handle_options'
	23: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/cli.rb:39:in `each'
	22: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/cli.rb:40:in `block in handle_options'
	21: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/cli.rb:92:in `block in pull'
	20: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/deployer/ssh.rb:48:in `pull_db'
	19: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/deployer/ssh/wpcli_sql_adapter.rb:34:in `adapt_remote_db!'
	18: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/deployer/ssh.rb:79:in `download_remote_db'
	17: from /var/lib/gems/2.7.0/gems/wordmove-5.2.1/lib/wordmove/deployer/ssh.rb:66:in `remote_run'
	16: from /var/lib/gems/2.7.0/gems/photocopier-1.4.0/lib/photocopier/ssh.rb:44:in `exec!'
	15: from /var/lib/gems/2.7.0/gems/photocopier-1.4.0/lib/photocopier/ssh.rb:71:in `session'
	14: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh.rb:255:in `start'
	13: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/session.rb:72:in `authenticate'
	12: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/session.rb:72:in `each'
	11: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/session.rb:86:in `block in authenticate'
	10: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/methods/publickey.rb:19:in `authenticate'
	 9: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/key_manager.rb:129:in `each_identity'
	 8: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/key_manager.rb:129:in `each'
	 7: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/key_manager.rb:132:in `block in each_identity'
	 6: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/methods/publickey.rb:20:in `block in authenticate'
	 5: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/methods/publickey.rb:62:in `authenticate_with'
	 4: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/key_manager.rb:167:in `sign'
	 3: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/key_factory.rb:42:in `load_private_key'
	 2: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/key_factory.rb:51:in `load_data_private_key'
	 1: from /var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/key_factory.rb:200:in `classify_key'
/var/lib/gems/2.7.0/gems/net-ssh-6.1.0/lib/net/ssh/authentication/ed25519_loader.rb:21:in `raiseUnlessLoaded': OpenSSH keys only supported if ED25519 is available (NotImplementedError)
net-ssh requires the following gems for ed25519 support:
 * ed25519 (>= 1.2, < 2.0)
 * bcrypt_pbkdf (>= 1.0, < 2.0)
See https://github.com/net-ssh/net-ssh/issues/565 for more information
Gem::MissingSpecError : "Could not find 'ed25519' (~> 1.2) among 70 total gem(s)
Checked in 'GEM_PATH=/home/preprod/.gem/ruby/2.7.0:/var/lib/gems/2.7.0:/usr/lib/ruby/gems/2.7.0:/usr/share/rubygems-integration/2.7.0:/usr/share/rubygems-integration/all:/usr/lib/x86_64-linux-gnu/rubygems-integration/2.7.0', execute `gem env` for more information"
```

https://github.com/net-ssh/net-ssh/blob/master/net-ssh.gemspec

I'm not sure what's the way to solve this. 

should I just install those gems ?

https://rubygems.org/gems/ed25519

https://www.rubydoc.info/gems/net-ssh

Thanks,

Eric

**Comment by alessandro-fazzi** [comment]
Just try `gem install ed25519`. `net-ssh` is already installed as a dependency for sure.

If it works let me know: probably we should add it to [photocopier](https://github.com/welaika/photocopier/)'s deps and I'd like to have feedback in order to decide what to do.

If it does not suffice, try also to read this discussion: https://github.com/welaika/wordmove/issues/586

UPDATE: other x-refs: https://github.com/welaika/docker-wordmove/issues/12 and https://github.com/welaika/wordmove/issues/510

**Reply by alessandro-fazzi** [reply]
@eldk I misunderstood what you intended about the gems but realized only after reading https://github.com/net-ssh/net-ssh README

```
gem install ed25519
gem install bcrypt_pbkdf
```

actually should be the way to go.

We, as maintainers, should eventually add boths as photocopier gem dependencies, or at least add some specific documentation here in the wiki.

**Reply by eldk** [reply]
Hello,

As wordmove is used on a ubuntu 20.04 server : 

I have to install ruby-dev : `apt-get install ruby-dev`

Then I installed ed25519 and bcrypt_pbkdf

```
gem install bcrypt_pbkdf
gem install ed25519
```
and all goes the way it should : database is dumped then pulled after been adapted for the local vhost :

```
-- Sending SELECT query...
-- Retrieving rows...
-- Disconnecting from localhost...

    local | gzip -9 -f "[secret]/wp-content/local-backup-1641853745.sql"

   remote | mysqldump --host=[secret] --user=naitupprod --password=[secret] --result-file="[secret]/wp-content/dump.sql" --verbose database

   remote | gzip -9 -f "[secret]/wp-content/dump.sql"

   remote | get: [secret]/wp-content/dump.sql.gz [secret]/wp-content/dump.sql.gz

   remote | delete: [secret]/wp-content/dump.sql.gz

    local | gzip -d -f "[secret]/wp-content/dump.sql.gz"

    local | mysql --host=[secret] --user=preprod --password=[secret] --database=database --execute="SET autocommit=0;SOURCE [secret]/wp-content/dump.sql;COMMIT"
mysql: [Warning] Using a password on the command line interface can be insecure.

    local | adapt dump for vhost

    local | wp search-replace --path=[secret] [secret] [secret] --quiet --skip-columns=guid --all-tables --allow-root

    local | adapt dump for wordpress_path

    local | wp search-replace --path=[secret] [secret]/ [secret] --quiet --skip-columns=guid --all-tables --allow-root

    local | delete: '[secret]/wp-content/dump.sql'
```

So that's OK/solved now. 

I don't know how to do it but maybe some check when pulling database with --simulate should alert ??

for example a simple `mysqldump --print-defaults` that will fail on those error type (ssh ...)

Thanks for that great tool,

Eric
